diff -Nur mysql-5.5.34-orig/plugin/auth/CMakeLists.txt mysql-5.5.34-patch/plugin/auth/CMakeLists.txt
--- mysql-5.5.34-orig/plugin/auth/CMakeLists.txt	2013-09-09 22:01:38.000000000 +0400
+++ mysql-5.5.34-patch/plugin/auth/CMakeLists.txt	2013-11-28 13:42:04.000000000 +0400
@@ -16,6 +16,8 @@
 
 MYSQL_ADD_PLUGIN(auth dialog.c 
   MODULE_ONLY)
+MYSQL_ADD_PLUGIN(dialog dialog.c 
+  MODULE_ONLY)
 MYSQL_ADD_PLUGIN(auth_test_plugin test_plugin.c 
   MODULE_ONLY)
 MYSQL_ADD_PLUGIN(qa_auth_interface qa_auth_interface.c
diff -Nur mysql-5.5.34-orig/plugin/auth/dialog.c mysql-5.5.34-patch/plugin/auth/dialog.c
--- mysql-5.5.34-orig/plugin/auth/dialog.c	2013-09-09 22:01:38.000000000 +0400
+++ mysql-5.5.34-patch/plugin/auth/dialog.c	2013-11-28 13:43:35.000000000 +0400
@@ -212,13 +212,26 @@
                          const char *prompt,
                          char *buf, int buf_len)
 {
-  char *ptr;
   fputs(prompt, stdout);
   fputc(' ', stdout);
-  if (fgets(buf, buf_len, stdin) == NULL)
-    return NULL;
-  if ((ptr= strchr(buf, '\n')))
-    *ptr= 0;
+
+  if (type == 2) /* password */
+  {
+    char *password;
+    password= getpass("");
+    strncpy(buf, password, buf_len);
+  }
+  else
+  {
+    if (!fgets(buf, buf_len-1, stdin))
+      buf[0]= 0;
+    else
+    {
+      int len= strlen(buf);
+      if (len && buf[len-1] == '\n')
+        buf[len-1]= 0;
+    }
+  }
 
   return buf;
 }
